### PR TITLE
Advantage Actor Critic (A2C) Model

### PR DESCRIPTION
**Pull request by blahBlahhhJ**
## What does this PR do?

Update for #596 (issue)
Implementation of A2C model for Reinforcement Learning
image

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? [not needed for typos/docs]
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-bolts/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by blahBlahhhJ** [comment]
@akihironitta 
Hi, I've finished the implementation for A2C, and tested the performance on OpenAI's CartPole environment (performance screenshot on the top). I think it's pretty good. However, I have a few questions regarding the PR.

1. In the 4th checklist item, does documentation means documentation in the code? If so, I think I wrote some amount of docs in my classes and methods.
2. In the 6th checklist item, regarding testing, I've wrote some tests for A2C, and all of them passed. But there are a few tests from other components (many from the data module section) that failed and I'm 100% sure I did not touch any of them. I assume the failure exist before I forked this repo. How do I deal with that?
3. For the last checklist item, I'm not sure if I should update the CHANGELOG, so I left it incomplete.

Let me know what I should do next. Thanks!

File: pl_bolts/models/rl/advantage_actor_critic_model.py [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+        if not isinstance(x, list):
+            x = [x]
+
+        if not isinstance(x, torch.Tensor):
+            x = torch.tensor(x, device=self.device)
+
+        logprobs, values = self.net(x)
+        return logprobs, values
+
+    def train_batch(self) -> Tuple[List[torch.Tensor], List[torch.Tensor], List[torch.Tensor]]:
```
Comment (akihironitta) [inline pl_bolts/models/rl/advantage_actor_critic_model.py:119]: The typing here seems conflicting with the docstring below. Would you mind fixing it?

File: pl_bolts/models/rl/advantage_actor_critic_model.py [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+            returns = self.compute_returns(self.batch_rewards, self.batch_masks, last_value)
+            for idx in range(self.hparams.batch_size):
+                yield self.batch_states[idx], self.batch_actions[idx], returns[idx]
+
+            self.batch_states = []
+            self.batch_actions = []
+            self.batch_rewards = []
+            self.batch_masks = []
+
+    def compute_returns(self, rewards, dones, last_value):
```
Comment (akihironitta) [inline pl_bolts/models/rl/advantage_actor_critic_model.py:163]: Could you add type hints here, too?

File: pl_bolts/models/rl/advantage_actor_critic_model.py [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+        for r, d in zip(rewards[::-1], dones[::-1]):
+            g = r + self.hparams.gamma * g * (1 - d)
+            returns.append(g)
+
+        # reverse list and stop the gradients
+        returns = torch.tensor(returns[::-1])
+
+        return returns
+
+    def loss(self, states, actions, returns):
```
Comment (akihironitta) [inline pl_bolts/models/rl/advantage_actor_critic_model.py:187]: Could you add type hints here, too?

**Review by akihironitta — COMMENTED** [review]
@blahBlahhhJ Hi, thank you for your contribution! The implementation looks great! I added a few commits directly to your branch to add its doc and to fix some minor issues. Also, I left some comments below, so would you mind having a look at them?

**Comment by akihironitta** [comment]
@blahBlahhhJ Thanks for the update!

1. The classes and methods were well documented! They were just not indexed in the docs, but I added them to the docs in 22f3b85. :)
2. You're right, the failing tests are irrelevant to the change in this PR. We will take care of that.
3. I updated the changelog for you.

File: pl_bolts/models/rl/advantage_actor_critic_model.py (hunk 1 above)
Comment (blahBlahhhJ) [reply pl_bolts/models/rl/advantage_actor_critic_model.py:119]: This method is actually a generator. The doctoring's note is the type for the DataLoader (stacking all the yield). I changed the typing to the actual type that is YIELDed from this generator (Tuple[np.ndarray, int, torch.Tensor]). Not sure if it's ok (not sure how typing works for generators)

File: docs/source/reinforce_learn.rst [hunk 4]
```diff
@@ -664,5 +647,17 @@ Example::
...
 
-.. autoclass:: pl_bolts.models.rl.vanilla_policy_gradient_model.VanillaPolicyGradient
+.. autoclass:: pl_bolts.models.rl.VanillaPolicyGradient
    :noindex:
+
+--------------
+
+Actor-Critic Models
+-------------------
+
```
[the remark this answers is not in the document]
Comment (blahBlahhhJ) [reply docs/source/reinforce_learn.rst]: Sure. I wrote a fair amount of documentation there, and uploaded an image (not sure if I'm allowed to do that). Let me know if there's any suggestion for the documentation.

[the remark this answers is not in the document]
Comment (blahBlahhhJ) [reply docs/source/reinforce_learn.rst]: Oh I also accidentally merged the master branch to my fork when trying to resolve a conflict in CHANGELOG. Not sure if that's a good practise...

[the remark this answers is not in the document]
Comment (akihironitta) [reply docs/source/reinforce_learn.rst]: Thank you for updating the docs. It looks great!

File: pl_bolts/models/rl/advantage_actor_critic_model.py (hunk 1 above)
Comment (akihironitta) [reply pl_bolts/models/rl/advantage_actor_critic_model.py]: Thanks for the update. For the record, I used `Iterator` as mentioned in the mypy docs: https://mypy.readthedocs.io/en/stable/kinds_of_types.html#generators.

**Comment by akihironitta** [comment]
@blahBlahhhJ Thank you for your work! Please allow us to have some more time to review the model.

**Comment by blahBlahhhJ** [comment]
> @blahBlahhhJ Thank you for your work! Please allow us to have some more time to review the model.

Any update for this PR? I would like to start implementing another algorithm (Soft Actor Critic) for this repo, and I think I could use some code in here. What would be a good practice to do this?

File: tests/models/rl/integration/test_actor_critic_models.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import argparse
+from unittest import TestCase
+
+import pytorch_lightning as pl
+
+from pl_bolts.models.rl.advantage_actor_critic_model import AdvantageActorCritic
+
+
+class TestActorCriticModels(TestCase):
```
Comment (akihironitta) [inline tests/models/rl/integration/test_actor_critic_models.py:9]: Could you use test functions instead of classes?

> pls do not deprecated unitest, rather a simple test functions

https://github.com/PyTorchLightning/lightning-bolts/pull/453/files#r615373664

**Review by akihironitta — COMMENTED** [review]
@blahBlahhhJ I'm sorry for the delay. Could you have a look at the comment?

> I think I could use some code in here. What would be a good practice to do this?

I think you can checkout from this branch to start implementing another one using the code in this PR.

@PyTorchLightning/core-bolts I'm not familiar with reinforcement learning, so could someone review this PR?

File: tests/models/rl/integration/test_actor_critic_models.py (hunk 5 above)
Comment (blahBlahhhJ) [reply tests/models/rl/integration/test_actor_critic_models.py:9]: sure! I'll look into that. Thanks for the update.

Comment (blahBlahhhJ) [reply tests/models/rl/integration/test_actor_critic_models.py]: Done. Also I passed a lot more tests (which is kind of strange).

**Comment by limberc** [comment]
Cool job!

I am actually working on this work too.

BTW,  can you add A3C too? Or can we do that?

**Comment by blahBlahhhJ** [comment]
> Cool job!
> 
> I am actually working on this work too.
> 
> BTW, can you add A3C too? Or can we do that?

Hi!

Currently I'm not planning to do A3C since it's more complicated and as I remembered from the paper, A2C has about the same performance as A3C. However, if you want to do a A3C, you could write an issue and then a pull request for it. It would be good if there's more RL methods in this repo I think.

**Comment by blahBlahhhJ** [comment]
@Borda 
Hi, Any update on this PR?
I think the code looks good. Could you review my code and merge this PR? Thanks in advance!

File: docs/source/reinforce_learn.rst [hunk 6]
```diff
@@ -21,10 +21,11 @@ Contributions by: `Donal Byrne <https://github.com/djbyrne>`_
 
 ------------
 
-.. note::
+.. note::=
```
Comment (Borda) [inline docs/source/reinforce_learn.rst:24]:
```suggestion
.. note::
```

File: pl_bolts/models/rl/advantage_actor_critic_model.py [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+    def train_dataloader(self) -> DataLoader:
+        """Get train loader"""
+        return self._dataloader()
+
+    def get_device(self, batch) -> str:
+        """Retrieve device currently being used by minibatch"""
+        return batch[0][0][0].device.index if self.on_gpu else "cpu"
+
+    @staticmethod
+    def add_model_specific_args(arg_parser: ArgumentParser) -> ArgumentParser:
```
Comment (Borda) [inline pl_bolts/models/rl/advantage_actor_critic_model.py:276]: Consider using `LightningCLI` see: https://devblog.pytorchlightning.ai/auto-structuring-deep-learning-projects-with-the-lightning-cli-9f40f1ef8b36

**Review by Borda — APPROVED** [review]

File: pl_bolts/models/rl/advantage_actor_critic_model.py (hunk 7 above)
Comment (blahBlahhhJ) [reply pl_bolts/models/rl/advantage_actor_critic_model.py:276]: Hi, I looked a bit into it, and It seems like in order to add the arguments (`add_model_specific_args()`) in the lightning module, I'll have to write my own subclass of `LightningCLI`, which defeats the purpose to use it to simplify the code. Let me know if my understanding is wrong. Or is there an example of using it in other RL algorithms?

Comment (Borda) [reply pl_bolts/models/rl/advantage_actor_critic_model.py:276]: the `LightningCLI` takes and maps what ever you have in Module init...
